cloud/parameterstore/parameter_record_db.go

```diff
@@ -0,0 +1,76 @@
+package parameterstore
+
+import (
+	"context"
+	"time"
+
+	"github.com/mongodb/anser/bsonutil"
+	adb "github.com/mongodb/anser/db"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// Collection holds metadata information about parameters.
+const Collection = "parameter_records"
+
+var (
+	nameKey        = bsonutil.MustHaveTag(ParameterRecord{}, "Name")
+	lastUpdatedKey = bsonutil.MustHaveTag(ParameterRecord{}, "LastUpdated")
+)
+
+// BumpParameterRecord bumps the parameter record to indicate that the parameter
+// was changed. information. This will not modify if the record exists and its
+// most recent update is more recent than lastUpdated.
+func BumpParameterRecord(ctx context.Context, db *mongo.Database, name string, lastUpdated time.Time) error {
+	res, err := db.Collection(Collection).UpdateOne(ctx, bson.M{
+		nameKey: name,
+		// Ensure that the latest parameter update wins (i.e. the update can
+		// only bump it forward in time and it can never go backwards).
+		lastUpdatedKey: bson.M{"$lt": lastUpdated},
+	}, bson.M{
+		"$set": bson.M{
+			lastUpdatedKey: lastUpdated,
+		},
+		"$setOnInsert": bson.M{
+			nameKey: name,
+		},
+	}, options.Update().SetUpsert(true))
+	if err != nil {
+		return err
+	}
+	if res.ModifiedCount == 0 && res.UpsertedCount == 0 {
+		return errors.Errorf("parameter record '%s' not upserted or modified", name)
```

Comment:
If it doesn't update it because the record in the db is newer, should that be considered an error?

Reply:
This intentionally returns an error so we can detect that the update did fail to go through, which can be used to detect races to modify the parameter. We can't actually do anything to fix this if this happens, but the caller will warn if this happens (and continue on error) once the follow-up PRs exist.